File: code.java
    // Complete the textQueries function below.
    static void textQueries(List<String> sentences, List<String> queries) {   
        List<HashMap<String, Integer>> sHash = new ArrayList<>();
        List<String[]> qArr = new ArrayList<>();
        
        //Cache query words in list of arrays
        for (int q = 0; q < queries.size(); q++) {
            qArr.add(queries.get(q).split(" "));
        }
        
        //Cache sentences words in a counter hashmap
        for (int s = 0; s < sentences.size(); s++) {
            HashMap<String, Integer> hash = new HashMap<>();
            for (String w : sentences.get(s).split(" ")) {
                if (hash.containsKey(w))
                    hash.put(w, hash.get(w)+1);
                else
                    hash.put(w, 1);
            }
            sHash.add(hash);
        }
        
        //Calculations and printing
        for (String[] q : qArr) {
            boolean found = false;

            for (int s = 0; s < sentences.size(); s++) {
                int lowest = 10; //tracks lowest count found word for printing

                for (String word : q) {
                    int count = 0;
                    
                    if (sHash.get(s).containsKey(word)) {
                        count = sHash.get(s).get(word);
                        if (count < lowest)
                            lowest = count;
                    }
                    else {
                        lowest = 0;
                        break;
                    }
                }
                
                for (int i = 0; i < lowest; i++) {
                    System.out.print(s + " ");
                    found = true;
                }

            }
            if (!found) 
                System.out.print("-1");
            System.out.println();
        }
    }

File: code.py
import math
import os
import random
import re
import sys
import collections
import time

def textQueries(sentences, queries):
    #code optimized by splitting operations reducing processing in triple loop
    
    sLen = len(sentences) # Number of sentences
    qLen = len(queries)   # Number of query strings
    sDict = {} # dict of sentence word counters {sentence#:[word1:count, word2:count]...}
    qDict = {} # dict of query word counters {query#:[word1:count, word2:count]...}
    
    # generate counter data
    for s in range(0, sLen):
        counter = collections.Counter()
        for w in sentences[s].split():
            counter[w] += 1
        sDict[s] = counter
    for q in range(0, qLen):
        counter = collections.Counter()
        for w in queries[q].split():
            counter[w] += 1
        qDict[q] = counter

    # process/print data   
    for q in range(0, qLen):
        found = False
        for s in range(0, sLen):
            count = 0
            rem = collections.Counter()
            rem = sDict[s].copy()
            
            while len(qDict[q]) is len(rem & qDict[q]):
                rem -= qDict[q]
                count += 1
            if count > 0:
                sys.stdout.write((str(s) + " ") * count)
                found = True
                
        if not found:
            print("-1")
        else:
            print("")
                             
    print("Time: ", time.clock(), "ms")




if __name__ == '__main__':
    sentences_count = int(input().strip())

    sentences = []

    for _ in range(sentences_count):
        sentences_item = input()
        sentences.append(sentences_item)

    queries_count = int(input().strip())

    queries = []

    for _ in range(queries_count):
        queries_item = input()
        queries.append(queries_item)

    textQueries(sentences, queries)
